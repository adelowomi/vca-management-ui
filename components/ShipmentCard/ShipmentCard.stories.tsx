import { select, text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';

import ShipmentCard from './ShipmentCard';

storiesOf('Shipment Card', module)
  .addDecorator(withKnobs)
  .add('Shipment Card Demo', () => (
    <ShipmentCard
      shippindId={text('Shipping ID', '26282922')}
      totalCapacity={text('Total Capacity', '29322')}
      status={select(
        'Status',
        {
          paid: 'paid',
          ongoing: 'ongoing',
          created: 'created',
        },
        'paid'
      )}
    />
  ));
