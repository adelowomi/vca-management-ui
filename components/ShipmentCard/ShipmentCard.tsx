/* eslint-disable @typescript-eslint/no-unused-vars */
import React, { useEffect, useMemo } from 'react';

import theme from '../../styles/theme';
import { HorizontalLine, MapPin, Origin } from '../AssetsSVG';

export const mapColorToStatus = {
  paid: {
    solid: theme.colors.darkishGreen,
    highlight: theme.colors.darkishGreenBkg,
  },
  ongoing: {
    solid: theme.colors.barneyPurple,
    highlight: theme.colors.barneyPurpleBkg,
  },
  created: {
    solid: theme.colors.blue,
    highlight: theme.colors.blueBkg,
  },
};

export type ShipmentStatusTypes = 'paid' | 'ongoing' | 'created';

interface Props {
  /**
   * ID of the shipment
   */
  shippindId: string;
  /**
   * Total capacity of the shipment
   */
  totalCapacity: string;
  /**
   * Status of the shipment
   * Can be either paid, ongoing, created
   */
  status: ShipmentStatusTypes;
  /**
   * Date which shipment departs
   */
  departureDate?: object;
  /**
   * Date which shipment arrivess
   */
  arrivalDate?: object;
}

const ShipmentCard = ({
  arrivalDate,
  departureDate,
  shippindId,
  status,
  totalCapacity,
}: Props) => {
  const statusColor = mapColorToStatus[status];
  useEffect(() => {
    if (statusColor) return;
    // eslint-disable-next-line no-console
    console.warn(
      `INVALID ${status} status in Shipment card. Status is wither 'paid', 'ongoing' or 'created'`
    );
  }, [statusColor, status]);

  const cardClassName = useMemo(() => {
    return {
      borderWidth: 0.7,
      borderLeftWidth: 5,
      borderLeftColor: statusColor && statusColor.solid,
    };
  }, [statusColor]);

  //TODO: Once we determine the formate and library to use for date and time
  // This function should return an object or construct the time
  // const contructTime = (date) => ()

  return (
    <div className="flex flex-row w-auto h-auto rounded" style={cardClassName}>
      <div className="flex items-start flex-row justify-between py-3 pr-16 pl-5 w-full">
        <div className="flex flex-col items-start">
          <p
            className="text-sm font-medium app-indigo mb-2"
            data-testid="shipmentId-test"
          >
            Shipment ID number #{shippindId}
          </p>
          <p className="text-xs app-greyish-brown mb-4">
            Total Capacity: {totalCapacity}
          </p>
          <p
            className="text-xs font-semibold py-2 px-3 rounded-2xl capitalize"
            data-testid="shipment-card-test"
            style={{
              color: statusColor && statusColor.solid,
              backgroundColor: statusColor && statusColor.highlight,
            }}
          >
            {status}
          </p>
        </div>
        <div>
          <div className="shipment-card-grid-1 mb-6">
            <div className="flex flex-row items-center">
              <Origin />
              <p className="text-base font-medium app-indigo ml-2">Origin</p>
            </div>
            <HorizontalLine className="m-auto" />
            <div className="flex flex-row items-center">
              <MapPin />
              <p className="text-base font-medium app-indigo ml-2">
                Destination
              </p>
            </div>
          </div>
          <div className="shipment-card-grid-2 mb-2">
            <p className="text-xs font-medium app-greyish-brown mb-2">
              Leaving: July 25th
            </p>
            <p className="text-xs font-medium app-greyish-brown mb-2">
              Estimated Arrival: 26th
            </p>
            <p className="text-xs font-semibold app-greyish-brown">7am -6pm</p>
            <p className="text-xs font-semibold app-greyish-brown">
              4pm to 8pm
            </p>
          </div>
        </div>
      </div>
    </div>
  );
};
export default ShipmentCard;
