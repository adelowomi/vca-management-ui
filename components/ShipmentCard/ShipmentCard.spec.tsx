import '@testing-library/jest-dom';

import { render } from '@testing-library/react';
import React from 'react';

// Components
import ShipmentCard, { ShipmentStatusTypes } from './ShipmentCard';

global.console.warn = jest.fn();

describe('ShipmentCard', () => {
  const status: ShipmentStatusTypes[] = ['paid', 'ongoing', 'created'];
  describe.each([status])('Given status is (%s)', (status) => {
    it('render status of shipment', () => {
      const { getByTestId, container } = render(
        <ShipmentCard shippindId={'125'} totalCapacity={'34'} status={status} />
      );
      expect(container.firstChild).toMatchSnapshot();
      expect(getByTestId('shipment-card-test')).toHaveTextContent(status);
    });
  });
  describe('Given invalid status is passed', () => {
    const originalError = console.error;
    const mockedError = jest.fn();
    beforeEach(() => {
      console.error = mockedError;
    });
    afterEach(() => {
      console.warn = originalError;
    });

    it('should call console.warn', () => {
      const status = 'unknown';
      render(
        // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
        // @ts-ignore *
        <ShipmentCard shippindId={'125'} totalCapacity={'34'} status={status} />
      );
      expect(console.warn).toBeCalledTimes(1);
      expect(console.warn).toHaveBeenLastCalledWith(
        `INVALID ${status} status in Shipment card. Status is wither 'paid', 'ongoing' or 'created'`
      );
    });
  });
});
