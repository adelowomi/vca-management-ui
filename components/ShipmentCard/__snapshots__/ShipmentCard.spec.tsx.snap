// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ShipmentCard Given status is (paid) render status of shipment 1`] = `
<div
  class="flex flex-row w-auto h-auto rounded"
  style="border-width: 0.7px; border-left-width: 5px; border-left-color: rgb(40, 118, 56);"
>
  <div
    class="flex items-start flex-row justify-between py-3 pr-16 pl-5 w-full"
  >
    <div
      class="flex flex-col items-start"
    >
      <p
        class="text-sm font-medium app-indigo mb-2"
        data-testid="shipmentId-test"
      >
        Shipment ID number #
        125
      </p>
      <p
        class="text-xs app-greyish-brown mb-4"
      >
        Total Capacity: 
        34
      </p>
      <p
        class="text-xs font-semibold py-2 px-3 rounded-2xl capitalize"
        data-testid="shipment-card-test"
        style="color: rgb(40, 118, 56); background-color: rgba(40, 118, 56, 0.1);"
      >
        paid
      </p>
    </div>
    <div>
      <div
        class="shipment-card-grid-1 mb-6"
      >
        <div
          class="flex flex-row items-center"
        >
          <svg
            height="7"
            viewBox="0 0 7 7"
            width="7"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3.5 0C5.433 0 7 1.52 7 3.397c0 1.876-1.567 3.397-3.5 3.397S0 5.274 0 3.397C0 1.521 1.567 0 3.5 0zm0 1.941c-.828 0-1.5.652-1.5 1.456 0 .804.672 1.456 1.5 1.456S5 4.2 5 3.397c0-.804-.672-1.456-1.5-1.456z"
              fill="#0C1A38"
            />
          </svg>
          <p
            class="text-base font-medium app-indigo ml-2"
          >
            Origin
          </p>
        </div>
        <svg
          class="m-auto"
          data-testid="horizontal-line"
          height="2"
          viewBox="0 0 105 2"
          width="105"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 1h106"
            fill="none"
            fill-rule="evenodd"
            stroke="#0C1A38"
            stroke-dasharray="3 3 3 3"
            stroke-linecap="round"
          />
        </svg>
        <div
          class="flex flex-row items-center"
        >
          <svg
            height="15"
            viewBox="0 0 11 15"
            width="11"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M5.5.441C2.467.441 0 2.618 0 5.294c0 .803.228 1.6.66 2.306l4.54 7.244c.06.096.175.156.3.156s.24-.06.3-.156l4.541-7.246c.431-.704.659-1.5.659-2.304C11 2.618 8.533.441 5.5.441zm0 7.28c-1.516 0-2.75-1.089-2.75-2.427S3.984 2.868 5.5 2.868s2.75 1.088 2.75 2.426S7.016 7.721 5.5 7.721z"
              fill="#0C1A38"
            />
          </svg>
          <p
            class="text-base font-medium app-indigo ml-2"
          >
            Destination
          </p>
        </div>
      </div>
      <div
        class="shipment-card-grid-2 mb-2"
      >
        <p
          class="text-xs font-medium app-greyish-brown mb-2"
        >
          Leaving: July 25th
        </p>
        <p
          class="text-xs font-medium app-greyish-brown mb-2"
        >
          Estimated Arrival: 26th
        </p>
        <p
          class="text-xs font-semibold app-greyish-brown"
        >
          7am -6pm
        </p>
        <p
          class="text-xs font-semibold app-greyish-brown"
        >
          4pm to 8pm
        </p>
      </div>
    </div>
  </div>
</div>
`;
